hazelcast/src/main/java/com/hazelcast/instance/ProtocolType.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.instance;
+
+import com.hazelcast.nio.Protocols;
+
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.Set;
+
+// todo move to internal package com.hazelcast.cluster.internal.impl ?
+
+/**
+ * Types of server sockets. A member typically listens on several types of server sockets
+ * for member-to-member, client-member protocol, WAN communication etc.
+ */
+public enum ProtocolType {
+
+    /**
+     * Ordinals of this enum are used for IDS inside {@link com.hazelcast.internal.cluster.impl.EndpointQualifier}
+     */
+    MEMBER(1, Protocols.CLUSTER),
+    CLIENT(1, Protocols.CLIENT_BINARY_NEW),
+    WAN(Integer.MAX_VALUE, Protocols.CLUSTER),
+    REST(1, Protocols.TEXT),
+    MEMCACHE(1, Protocols.TEXT);
+
+    private static final Set<ProtocolType> ALL_PROTOCOL_TYPES;
+
+    static {
+        Set<ProtocolType> allProtocolTypes = EnumSet.allOf(ProtocolType.class);
+        ALL_PROTOCOL_TYPES = Collections.unmodifiableSet(allProtocolTypes);
+    }
+
+    private final int serverSocketCardinality;
+    private final String descriptor;
+
+    ProtocolType(int serverSocketCardinality, String descriptor) {
+        this.serverSocketCardinality = serverSocketCardinality;
+        this.descriptor = descriptor;
+    }
+
+    public static ProtocolType valueOf(int ordinal) {
+        return ProtocolType.values()[ordinal];
```

Comment:
If endpoints are frequently deserialized, it might be a good idea to cache `ProtocolType.values()` to avoid creating a new array on every invocation.